Make test case creation code more compatible with meta tensors.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74313

The incompatibilities involve:

- A lot of code allocates a tensor on device, and then immediately
  reads it out item() to get a Python value.  This doesn't work
  for meta tensors, so I make sure to allocate the tensor on CPU
  in these cases. (Arguably, it would be better if CUDA tests also
  allocated these scalar values on CPU; if you really want me to do
  this as part of this change shout.)
- A common pattern in linear algebra generation code is to generate
  some input, and then iterate over the matrix looking for certain
  numbers and "fixing" them up in some way.  This process doesn't work
  for meta tensors, so I skip any extra initialization when teh
  generated input is meta.

This commit doesn't directly add any tests as I have to add a few
more fixes (later in this stack) before I am able to run all OpInfo
sample input generation commands on meta tensors.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D34931337](https://our.internmc.facebook.com/intern/diff/D34931337)

ezyang committed:
Make test case creation code more compatible with meta tensors.

The incompatibilities involve:

- A lot of code allocates a tensor on device, and then immediately
  reads it out item() to get a Python value.  This doesn't work
  for meta tensors, so I make sure to allocate the tensor on CPU
  in these cases. (Arguably, it would be better if CUDA tests also
  allocated these scalar values on CPU; if you really want me to do
  this as part of this change shout.)
- A common pattern in linear algebra generation code is to generate
  some input, and then iterate over the matrix looking for certain
  numbers and "fixing" them up in some way.  This process doesn't work
  for meta tensors, so I skip any extra initialization when teh
  generated input is meta.

This commit doesn't directly add any tests as I have to add a few
more fixes (later in this stack) before I am able to run all OpInfo
sample input generation commands on meta tensors.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34931337).

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -3457,11 +3461,12 @@ def large_1d_unique():
 
 def sample_inputs_threshold(op_info, device, dtype, requires_grad, **kwargs):
     make_arg = partial(make_tensor, dtype=dtype, device=device, requires_grad=requires_grad)
+    make_scalar_arg = partial(make_tensor, dtype=dtype, device='cpu', requires_grad=requires_grad)
```
Nit: this could pass the correct shape and take the `.item()` here so you wouldn't need to repeat yourself later

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2012,7 +2012,10 @@ def sample_inputs_binary_pwise(
     )
 
-    scalar = make_arg((), **rhs_make_tensor_kwargs)
+    scalar_kwargs = copy.copy(rhs_make_tensor_kwargs)
```
Note this will be a merge conflict with https://github.com/pytorch/pytorch/pull/74280, so whoever merges second will need to update their PR

albanD commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -6522,10 +6532,11 @@ def sample_inputs_rsub(op_info, device, dtype, requires_grad, other_scalar, **kw
         if other_scalar:
-            y = y.item()
-        a = make_arg((), dtype=dtype_a).item()
+            y = make_tensor(shape_y, device='cpu', dtype=dtype_y, requires_grad=requires_grad_y).item()
```
Sounds like it is a bug to set `requires_grad_y=True` here no?

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2487,11 +2487,12 @@ def random_square_matrix_of_rank(l, rank, dtype=torch.double, device='cpu'):
-            s[i] = 0
-        elif s[i] == 0:
-            s[i] = 1
+    if not s.is_meta:
```
Comment here would be nice

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -4808,8 +4814,8 @@ def make_nonzero_det(A, *, sign=1, min_singular_value=0.1, **kwargs):
         A = (u * s.unsqueeze(-2)) @ vh
-        det = A.det()
-        if sign is not None:
+        if sign is not None and not A.is_meta:
```
Comment here would be cool

mruberry reviewed (approved):
Cool!

Bonus points if you update 

https://github.com/pytorch/pytorch/blob/e084e266da261fb25742cb4caa60ce9e79181e3a/torch/testing/_internal/common_methods_invocations.py#L327

with some tips for handling meta tensors appropriately (like the best practice for generating scalar arguments)

mruberry commented:
Follow-up question... should we add `make_scalar` to the torch.testing API? cc @pmeier

ezyang commented on torch/testing/_creation.py:
```diff
@@ -139,10 +139,18 @@ def clamp(a, l, h):
+        if result.is_meta:
+            # TODO: repeat_interleave is a bad composite for meta
+            # see https://github.com/pytorch/pytorch/issues/74300
+            result_size = list(result.size())
```
This implementation is wrong for non-meta, it doesn't actually preserve the data that was created. Though I guess we could `copy_` it in after the fact.

albanD replied:
Not a big deal. Just avoiding the if would be nice.

ezyang committed:
Update on "Make test case creation code more compatible with meta tensors."


The incompatibilities involve:

- A lot of code allocates a tensor on device, and then immediately
  reads it out item() to get a Python value.  This doesn't work
  for meta tensors, so I make sure to allocate the tensor on CPU
  in these cases. (Arguably, it would be better if CUDA tests also
  allocated these scalar values on CPU; if you really want me to do
  this as part of this change shout.)
- A common pattern in linear algebra generation code is to generate
  some input, and then iterate over the matrix looking for certain
  numbers and "fixing" them up in some way.  This process doesn't work
  for meta tensors, so I skip any extra initialization when teh
  generated input is meta.

This commit doesn't directly add any tests as I have to add a few
more fixes (later in this stack) before I am able to run all OpInfo
sample input generation commands on meta tensors.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

Differential Revision: [D34931337](https://our.internmc.facebook.com/intern/diff/D34931337)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34931337).

pmeier commented:
> Follow-up question... should we add `make_scalar` to the torch.testing API?

IMO, no. It is a trivial parametrization of `make_tensor`:

```py
make_scalar = partial(make_tensor, ())
```

Similar to `assert_equal = partial(assert_close, rtol=0, atol=0)`, we could add a note in the documentation.

ezyang committed:
Update on "Make test case creation code more compatible with meta tensors."


The incompatibilities involve:

- A lot of code allocates a tensor on device, and then immediately
  reads it out item() to get a Python value.  This doesn't work
  for meta tensors, so I make sure to allocate the tensor on CPU
  in these cases. (Arguably, it would be better if CUDA tests also
  allocated these scalar values on CPU; if you really want me to do
  this as part of this change shout.)
- A common pattern in linear algebra generation code is to generate
  some input, and then iterate over the matrix looking for certain
  numbers and "fixing" them up in some way.  This process doesn't work
  for meta tensors, so I skip any extra initialization when teh
  generated input is meta.

This commit doesn't directly add any tests as I have to add a few
more fixes (later in this stack) before I am able to run all OpInfo
sample input generation commands on meta tensors.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

Differential Revision: [D34931337](https://our.internmc.facebook.com/intern/diff/D34931337)

[ghstack-poisoned]

ezyang commented:
@pmeier I'm not going to add it to the API because I'm lazy, but I don't think I agree with you here? The parametrization is nontrivial because you have to remember to set device='cpu' and you call item() on the result.

ezyang commented:
> with some tips for handling meta tensors appropriately (like the best practice for generating scalar arguments)

There's no logical place to put these comments, `make_tensor` isn't mentioned at all haha

ezyang committed:
Update on "Make test case creation code more compatible with meta tensors."


The incompatibilities involve:

- A lot of code allocates a tensor on device, and then immediately
  reads it out item() to get a Python value.  This doesn't work
  for meta tensors, so I make sure to allocate the tensor on CPU
  in these cases. (Arguably, it would be better if CUDA tests also
  allocated these scalar values on CPU; if you really want me to do
  this as part of this change shout.)
- A common pattern in linear algebra generation code is to generate
  some input, and then iterate over the matrix looking for certain
  numbers and "fixing" them up in some way.  This process doesn't work
  for meta tensors, so I skip any extra initialization when teh
  generated input is meta.

This commit doesn't directly add any tests as I have to add a few
more fixes (later in this stack) before I am able to run all OpInfo
sample input generation commands on meta tensors.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

Differential Revision: [D34931337](https://our.internmc.facebook.com/intern/diff/D34931337)

[ghstack-poisoned]

ezyang committed:
Update on "Make test case creation code more compatible with meta tensors."


The incompatibilities involve:

- A lot of code allocates a tensor on device, and then immediately
  reads it out item() to get a Python value.  This doesn't work
  for meta tensors, so I make sure to allocate the tensor on CPU
  in these cases. (Arguably, it would be better if CUDA tests also
  allocated these scalar values on CPU; if you really want me to do
  this as part of this change shout.)
- A common pattern in linear algebra generation code is to generate
  some input, and then iterate over the matrix looking for certain
  numbers and "fixing" them up in some way.  This process doesn't work
  for meta tensors, so I skip any extra initialization when teh
  generated input is meta.

This commit doesn't directly add any tests as I have to add a few
more fixes (later in this stack) before I am able to run all OpInfo
sample input generation commands on meta tensors.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

Differential Revision: [D34931337](https://our.internmc.facebook.com/intern/diff/D34931337)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34931337).

pmeier commented:
> The parametrization is nontrivial because you have to remember to set device='cpu' and you call item() on the result.

Ah, yeah, I missed that we were talking about Python scalars rather than tensor scalars, i.e. 0d tensors. In that case I agree it needs to be a thin wrapper. Still not sure if we want something like that for the public API.

mruberry commented:
> > with some tips for handling meta tensors appropriately (like the best practice for generating scalar arguments)
> 
> There's no logical place to put these comments, `make_tensor` isn't mentioned at all haha

Doh! That's OK, we should just add `make_scalar` and be done with it.

mruberry commented:
I filed issue https://github.com/pytorch/pytorch/issues/74531 for make_scalar.

ezyang committed:
Update on "Make test case creation code more compatible with meta tensors."


The incompatibilities involve:

- A lot of code allocates a tensor on device, and then immediately
  reads it out item() to get a Python value.  This doesn't work
  for meta tensors, so I make sure to allocate the tensor on CPU
  in these cases. (Arguably, it would be better if CUDA tests also
  allocated these scalar values on CPU; if you really want me to do
  this as part of this change shout.)
- A common pattern in linear algebra generation code is to generate
  some input, and then iterate over the matrix looking for certain
  numbers and "fixing" them up in some way.  This process doesn't work
  for meta tensors, so I skip any extra initialization when teh
  generated input is meta.

This commit doesn't directly add any tests as I have to add a few
more fixes (later in this stack) before I am able to run all OpInfo
sample input generation commands on meta tensors.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

Differential Revision: [D34931337](https://our.internmc.facebook.com/intern/diff/D34931337)

[ghstack-poisoned]
